tests/problems/generic/test_gw_problem.py

```diff
@@ -123,18 +131,38 @@ def test_pass_arguments(self, adata_space_rotate: AnnData, args_to_check: Mappin
             assert getattr(geom, val) == args_to_check[arg]
 
     @pytest.mark.fast()
-    @pytest.mark.parametrize("cost", [("sq_euclidean", SqEuclidean), ("euclidean", Euclidean), ("cosine", Cosine)])
-    def test_prepare_costs(self, adata_time: AnnData, cost: Tuple[str, Any]):
+    @pytest.mark.parametrize(
+        ("cost_str", "cost_inst", "cost_kwargs"),
+        [
+            ("sq_euclidean", SqEuclidean, {}),
+            ("euclidean", Euclidean, {}),
+            ("cosine", Cosine, {}),
+            ("pnorm_p", PNormP, {"p": 3}),
+            ("sq_pnorm", SqPNorm, {"x": {"p": 3}, "y": {"p": 4}}),
+            ("elastic_l1", ElasticL1, {"x": {"gamma": 3}, "y": {"gamma": 4}}),
+            ("elastic_stvs", ElasticSTVS, {"x": {"gamma": 3}, "y": {"gamma": 4}}),
+        ],
+    )
+    def test_prepare_costs(self, adata_time: AnnData, cost_str: str, cost_inst: Any, cost_kwargs: Mapping[str, int]):
         problem = GWProblem(adata=adata_time)
         problem = problem.prepare(
-            key="time",
-            policy="sequential",
-            x_attr="X_pca",
-            y_attr="X_pca",
-            cost=cost[0],
+            key="time", policy="sequential", x_attr="X_pca", y_attr="X_pca", cost=cost_str, cost_kwargs=cost_kwargs
         )
-        assert isinstance(problem[0, 1].x.cost, cost[1])
-        assert isinstance(problem[0, 1].y.cost, cost[1])
+        assert isinstance(problem[0, 1].x.cost, cost_inst)
+        assert isinstance(problem[0, 1].y.cost, cost_inst)
+
+        if cost_kwargs:
+            if len(cost_kwargs) > 1:
+                k, v = next(iter(cost_kwargs["x"].items()))
```

Comment:
This again iterates only for 1 parameter of the cost, but some costs can (and will) have more; please use `loop`, just like in the `handle_cost` case.

Reply:
sorry, my bad!